c/secp256k1_ripemd160_sha256_sighash_all.c

```diff
@@ -0,0 +1,190 @@
+  flatbuffers_uint8_vec_t seq = ns(Bytes_seq(bytes));
+  size_t len = flatbuffers_uint8_vec_len(seq);
+
+  if (len > *s) {
+    return ERROR_BUFFER_NOT_ENOUGH;
+  }
+
+  for (size_t i = 0; i < len; i++) {
+    buffer[i] = flatbuffers_uint8_vec_at(seq, i);
+  }
+  *s = len;
+
+  return CKB_SUCCESS;
+}
+
+/*
+ * Arguments are listed in the following order:
+ * 0. program name
+ * 1. ripemd160(sha256(pubkey)) hash, used to
+ * shield the real pubkey in lock script.
+ *
+ * Witness:
+ * 3. signature, signature used to present ownership
+ * 4. signature size, in little endian 64 bit unsigned integer
+ * 5. pubkey, real pubkey used to identify token owner
+ */
+int main(int argc, char* argv[]) {
+  int ret;
+  size_t index = 0;
+  uint64_t signature_size = 0;
+  volatile uint64_t len = 0;
+  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
+  unsigned char temp[TEMP_SIZE];
+  unsigned char witness[WITNESS_SIZE];
+  ns(Witness_table_t) witness_table;
+  ns(Bytes_vec_t) args;
+  uint8_t secp_data[CKB_SECP256K1_DATA_SIZE];
+
+  if (argc != 2) {
+    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  }
+
+  secp256k1_context context;
+  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
+  if (ret != 0) {
+    return ret;
+  }
+
+  len = BLAKE2B_BLOCK_SIZE;
+  ret = ckb_load_tx_hash(tx_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  while (1) {
+    len = 0;
+    /*
+     * Actually we don't need this syscall, we are just making it to grab all
+     * input indices for current group, from which we can load the actual
+     * witness data we need. `since` field is chosen here since it has a fixed
+     * size of 8 bytes, which is both predictable, and also provides minimal
+     * cycle consumption.
+     */
+    ret = ckb_load_input_by_field(NULL, &len, 0, index, CKB_SOURCE_GROUP_INPUT,
+                                  CKB_INPUT_FIELD_SINCE);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      return 0;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+
+    /* Now we load actual witness data using the same input index above. */
+    len = WITNESS_SIZE;
+    ret = ckb_load_witness(witness, &len, 0, index, CKB_SOURCE_GROUP_INPUT);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_SYSCALL;
+    }
+    if (len > WITNESS_SIZE) {
+      return ERROR_WITNESS_TOO_LONG;
+    }
+
+    if (!(witness_table = ns(Witness_as_root(witness)))) {
+      return ERROR_ENCODING;
+    }
+    args = ns(Witness_data(witness_table));
+    if (ns(Bytes_vec_len(args)) < 2) {
+      return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+    }
+
+    /* check pubkey hash */
+    len = TEMP_SIZE;
+    ret = extract_bytes(ns(Bytes_vec_at(args, 1)), temp, &len);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_ENCODING;
+    }
+    sha256_state sha256_ctx;
+    ripemd160_state ripe160_ctx;
+
+    sha256_init(&sha256_ctx);
+    sha256_update(&sha256_ctx, temp, len);
+    sha256_finalize(&sha256_ctx, temp);
```

Comment:
We can set up a buffer to hold the hash result for sha256 and ripemd160. Thus we don't need to read the public key again below.

Reply:
Or we can reorder the code. Parse the pubkey after it is extraced from witness.